File: sdks/typescript/src/core/dual-hash.ts
/**
 * DCP v2.0 Dual-Hash infrastructure (SHA-256 + SHA3-256).
 *
 * Running two independent hash families from day one provides defense-in-depth:
 * SHA-2 (Merkle-Damgård) and SHA-3 (Keccak sponge) use completely different
 * internal constructions, so a break in one is unlikely to affect the other.
 */

import { createHash } from 'crypto';

export interface DualHash {
  sha256: string;
  sha3_256: string;
}

export function sha256Hex(data: Uint8Array | Buffer | string): string {
  return createHash('sha256').update(data).digest('hex');
}

export function sha3_256Hex(data: Uint8Array | Buffer | string): string {
  return createHash('sha3-256').update(data).digest('hex');
}

/** Compute both SHA-256 and SHA3-256 over the same input. */
export function dualHash(data: Uint8Array | Buffer | string): DualHash {
  return {
    sha256: sha256Hex(data),
    sha3_256: sha3_256Hex(data),
  };
}

/** Dual-hash a canonicalized JSON payload (as UTF-8 bytes). */
export function dualHashCanonical(canonicalJson: string): DualHash {
  const bytes = Buffer.from(canonicalJson, 'utf8');
  return dualHash(bytes);
}

/** Compute dual Merkle roots from hex leaf hashes. */
export function dualMerkleRoot(leaves: DualHash[]): DualHash | null {
  if (!leaves || leaves.length === 0) return null;

  let sha2Layer = leaves.map((l) => l.sha256);
  let sha3Layer = leaves.map((l) => l.sha3_256);

  sha2Layer = merkleReduce(sha2Layer, 'sha256');
  sha3Layer = merkleReduce(sha3Layer, 'sha3-256');

  return { sha256: sha2Layer[0], sha3_256: sha3Layer[0] };
}

function merkleReduce(leaves: string[], hashAlg: string): string[] {
  let layer = leaves.slice();
  while (layer.length > 1) {
    if (layer.length % 2 === 1) layer.push(layer[layer.length - 1]);
    const next: string[] = [];
    for (let i = 0; i < layer.length; i += 2) {
      const left = Buffer.from(layer[i], 'hex');
      const right = Buffer.from(layer[i + 1], 'hex');
      next.push(
        createHash(hashAlg)
          .update(Buffer.concat([left, right]))
          .digest('hex'),
      );
    }
    layer = next;
  }
  return layer;
}

File: sdks/python/dcp_ai/v2/dual_hash.py
"""
Dual-hash (SHA-256 + SHA3-256) utilities for DCP v2 post-quantum readiness.
"""

from __future__ import annotations

import hashlib


def sha256_hex(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def sha3_256_hex(data: bytes) -> str:
    return hashlib.sha3_256(data).hexdigest()


def dual_hash(data: bytes) -> dict[str, str]:
    return {
        "sha256": sha256_hex(data),
        "sha3_256": sha3_256_hex(data),
    }


def dual_hash_canonical(canonical_json: str) -> dict[str, str]:
    return dual_hash(canonical_json.encode("utf-8"))


def _merkle_root_hex(leaves: list[str], hash_fn: type[object] | None = None) -> str | None:
    """Compute a Merkle root from hex-encoded leaf hashes using the given hashlib constructor."""
    if not leaves:
        return None
    layer = list(leaves)
    while len(layer) > 1:
        if len(layer) % 2 == 1:
            layer.append(layer[-1])
        next_layer: list[str] = []
        for i in range(0, len(layer), 2):
            combined = bytes.fromhex(layer[i]) + bytes.fromhex(layer[i + 1])
            if hash_fn is not None:
                next_layer.append(hash_fn(combined).hexdigest())  # type: ignore[operator]
            else:
                next_layer.append(hashlib.sha256(combined).hexdigest())
        layer = next_layer
    return layer[0]


def dual_merkle_root(leaves: list[dict[str, str]]) -> dict[str, str] | None:
    """Compute dual Merkle roots from a list of dual-hash dicts.

    Each leaf dict must have 'sha256' and 'sha3_256' keys.
    Returns dict with sha256 and sha3_256 Merkle roots, or None if empty.
    """
    if not leaves:
        return None
    sha256_leaves = [leaf["sha256"] for leaf in leaves]
    sha3_leaves = [leaf["sha3_256"] for leaf in leaves]
    return {
        "sha256": _merkle_root_hex(sha256_leaves, hashlib.sha256) or "",
        "sha3_256": _merkle_root_hex(sha3_leaves, hashlib.sha3_256) or "",
    }
